### PR TITLE
En 6312/meta tx standard

### PR DESCRIPTION
**Pull request by sasurobert**
Implemented meta transaction standard.

Attached documentation: https://docs.google.com/document/d/1P-ZNs2BCvHheb9jAkVC6MYWsWfGoQzAHbvBnrvezXAQ/edit

### CODE REVIEW

File: process/transaction/interceptedTransaction.go [hunk 1]
```diff
@@ -137,6 +148,55 @@ func (inTx *InterceptedTransaction) CheckValidity() error {
 	return nil
 }
 
+func (inTx *InterceptedTransaction) verifyIfRelayedTx(tx *transaction.Transaction) error {
```
Comment (iulianpascalau) [inline process/transaction/interceptedTransaction.go:151]: Tests missing for this function

File: process/transaction/shardProcess.go [hunk 2]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
 	return nil
 }
 
+func (txProc *txProcessor) processRelayedTx(
```
Comment (iulianpascalau) [inline process/transaction/shardProcess.go:412]: This needs heavy testing

File: process/smartContract/process.go [hunk 3]
```diff
@@ -279,7 +285,7 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
 	acntDst, err = sc.reloadLocalAccount(acntDst)
 	if err != nil {
 		log.Debug("reloadLocalAccount error", "error", err.Error())
-		return nil
+		return 0, err
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:288]: when critical error is returned the return code is not used, like sending nil back when critical error

File: process/transaction/interceptedTransaction.go (hunk 1 above)
Comment (sasurobert) [reply process/transaction/interceptedTransaction.go:151]: agreed after first reviews will do it. may merge after disabling this txtype

File: process/transaction/shardProcess.go (hunk 2 above)
Comment (sasurobert) [reply process/transaction/shardProcess.go:412]: agreed after first reviews will do it. may merge after disabling.

File: process/smartContract/process.go [hunk 4]
```diff
@@ -421,27 +429,117 @@ func (sc *scProcessor) ProcessIfError(
...
+	return nil
+}
+
+func isRelayedTx(tx data.TransactionHandler) (*smartContractResult.SmartContractResult, bool) {
+	relayedSCR, ok := tx.(*smartContractResult.SmartContractResult)
+	if !ok {
+		return nil, false
+	}
+
+	if len(relayedSCR.RelayerAddr) == len(relayedSCR.SndAddr) &&
```
Comment (andreibancioiu) [inline process/smartContract/process.go:507]: Perhaps add a comment to explain the reason behind this condition?

File: process/smartContract/process.go [hunk 5]
```diff
@@ -421,27 +429,117 @@ func (sc *scProcessor) ProcessIfError(
...
+		relayedSCR.RelayedValue != nil && relayedSCR.RelayedValue.Cmp(zero) > 0 {
+		return relayedSCR, true
+	}
+
+	return nil, false
+}
+
+func (sc *scProcessor) addBackTxValues(
```
Comment (andreibancioiu) [inline process/smartContract/process.go:515]: Perhaps add a short comment to explain the purpose of the function?

File: process/smartContract/process.go [hunk 6]
```diff
@@ -1050,18 +1149,28 @@ func (sc *scProcessor) getAccountFromAddress(address []byte) (state.UserAccountH
...
 
 	log.Trace("scProcessor.ProcessSmartContractResult()", "sender", scr.GetSndAddr(), "receiver", scr.GetRcvAddr())
 
 	var err error
+	returnCode := vmcommon.UserError
 	txHash, err := core.CalculateHash(sc.marshalizer, sc.hasher, scr)
 	if err != nil {
 		log.Debug("CalculateHash error", "error", err)
-		return err
+		return returnCode, err
```
Comment (andreibancioiu) [inline process/smartContract/process.go:1164]: Isn't this a critical error?

File: process/smartContract/argumentParser.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
+package smartContract
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/argumentParser.go:1]: agreed. but this way, there was less changes to do in the code :D

File: process/transaction/shardProcess.go [hunk 8]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+	relayerFee := big.NewInt(0).Mul(big.NewInt(0).SetUint64(relayerGasLimit), big.NewInt(0).SetUint64(tx.GetGasPrice()))
+	totalFee := big.NewInt(0).Mul(big.NewInt(0).SetUint64(tx.GetGasLimit()), big.NewInt(0).SetUint64(tx.GetGasPrice()))
+	remainingFee := big.NewInt(0).Sub(totalFee, relayerFee)
+
+	txHash, err := core.CalculateHash(txProc.marshalizer, txProc.hasher, tx)
+	if err != nil {
+		return 0, err
+	}
+
+	if !check.IfNil(relayerAcnt) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:452]: no. relayer can be in another shard. and that this was a cross shard transaction.

File: process/transaction/shardProcess.go [hunk 9]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+	if check.IfNil(acntDst) {
+		return vmcommon.Ok, nil
+	}
+
+	err = acntDst.AddToBalance(tx.GetValue())
+	if err != nil {
+		return 0, err
+	}
+
+	err = acntDst.AddToBalance(remainingFee)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:481]: relayer pays for all the fee and tx.Value from the Big transaction. The remaining FEE + tx.Value is sent to the USER. From there the user transaction is executed.

File: process/transaction/shardProcess.go [hunk 10]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+	}
+
+	return txProc.processUserTx(userTx, adrSrc, tx.Value, tx.Nonce, txHash)
+}
+
+func (txProc *txProcessor) processUserTx(
+	tx *transaction.Transaction,
+	relayerAdr []byte,
+	relayedTxValue *big.Int,
+	relayedNonce uint64,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:498]: Yes. it is correct :).

File: process/transaction/shardProcess.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:412]: yes. If you have any other good naming solution, we can change it.

File: process/transaction/shardProcess.go [hunk 11]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+	errorMsg string,
+) error {
+	userAcnt, err := txProc.getAccountFromAddress(userAdr)
+	if err != nil {
+		return err
+	}
+	if check.IfNil(userAcnt) {
+		return process.ErrNilAccountsAdapter
+	}
+	err = userAcnt.SubFromBalance(relayedTxValue)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:601]: ErrLinAccounts adapter will not happen - as the function is called only if userAcnt is not nil. This error is from a critical coding error.

File: process/transaction/shardProcess.go [hunk 12]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+		OriginalTxHash: originalTxHash,
+		ReturnMessage:  []byte(errorMsg),
+	}
+
+	relayerAcnt, err := txProc.getAccountFromAddress(relayerAdr)
+	if err != nil {
+		return err
+	}
+
+	if !check.IfNil(relayerAcnt) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:620]: yes. It is possible. That way you resolve the payback through SCR.

File: process/smartContract/process.go (hunk 6 above)
Comment (sasurobert) [reply process/smartContract/process.go:1164]: this is a critical error.

File: process/transaction/shardProcess.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/transaction/shardProcess.go:481]: Makes sense.

File: go.sum [hunk 13]
```diff
@@ -18,8 +20,11 @@ github.com/ElrondNetwork/elrond-vm-common v0.0.0-20191203115206-691b00a6e76a/go.
 github.com/ElrondNetwork/elrond-vm-common v0.1.9/go.mod h1:ZakxPST/Wt8umnRtA9gobcy3Dw2bywxwkC54P5VhO9g=
 github.com/ElrondNetwork/elrond-vm-common v0.1.19 h1:mRO768HtMyXY23pvG18DonKVEIlNvXxoyKP94S9fb2A=
 github.com/ElrondNetwork/elrond-vm-common v0.1.19/go.mod h1:ZakxPST/Wt8umnRtA9gobcy3Dw2bywxwkC54P5VhO9g=
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply go.sum:22]: done

File: core/constants.go [hunk 14]
```diff
@@ -414,6 +414,9 @@ const BuiltInFunctionSaveKeyValue = "SaveKeyValue"
 // BuiltInFunctionESDTTransfer is the key for the elrond standard digital token transfer built-in function
 const BuiltInFunctionESDTTransfer = "ESDTTransfer"
 
+// RelayedTransaction is the key for the elrond meta/gassless/relayed transaction standard
+const RelayedTransaction = "relayedTx"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/constants.go:418]: done.

File: process/errors.go [hunk 15]
```diff
@@ -790,3 +787,15 @@ var ErrOriginatorIsBlacklisted = errors.New("originator is blacklisted")
...
+var ErrInvalidAddressInRelayedTx = errors.New("invalid address in relayed tx")
+
+// ErrInvalidVMType signals that invalid vm type was provided
+var ErrInvalidVMType = errors.New("invalid VM type")
+
+// ErrRecursiveRelayedTXIsNotAllowed signals that recursive relayed tx is not allowed
+var ErrRecursiveRelayedTXIsNotAllowed = errors.New("recursive relayed tx is not allowed")
+
+// ErrRelayedTxValueHigherThenUserTxValue signals that relayed tx value is higher then user tx value
+var ErrRelayedTxValueHigherThenUserTxValue = errors.New("relayed tx value is higher then user tx value")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/errors.go:801]: done

File: process/errors.go [hunk 16]
```diff
@@ -790,3 +787,15 @@ var ErrOriginatorIsBlacklisted = errors.New("originator is blacklisted")
...
 var ErrShardIsStuck = errors.New("shard is stuck")
+
+// ErrInvalidAddressInRelayedTx signals that an invalid address was provided in the relayed tx
+var ErrInvalidAddressInRelayedTx = errors.New("invalid address in relayed tx")
+
+// ErrInvalidVMType signals that invalid vm type was provided
+var ErrInvalidVMType = errors.New("invalid VM type")
+
+// ErrRecursiveRelayedTXIsNotAllowed signals that recursive relayed tx is not allowed
+var ErrRecursiveRelayedTXIsNotAllowed = errors.New("recursive relayed tx is not allowed")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/errors.go:798]: done

File: process/transaction/interceptedTransaction.go [hunk 17]
```diff
@@ -137,6 +148,55 @@ func (inTx *InterceptedTransaction) CheckValidity() error {
...
+		return process.ErrInvalidArguments
+	}
+
+	userTx, err := createTx(inTx.protoMarshalizer, userTxArgs[0])
+	if err != nil {
+		return err
+	}
+
+	if !bytes.Equal(userTx.SndAddr, tx.RcvAddr) {
+		return process.ErrInvalidAddressInRelayedTx
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/interceptedTransaction.go:170]: it is explained in the error. But changed.

File: process/transaction/interceptedTransaction.go [hunk 18]
```diff
@@ -137,6 +148,55 @@ func (inTx *InterceptedTransaction) CheckValidity() error {
...
+	if err != nil {
+		return err
+	}
+
+	err = inTx.verifySig(userTx)
+	if err != nil {
+		return err
+	}
+
+	// recursive relayed transactions are not allowed
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/interceptedTransaction.go:183]: done

File: process/transaction/interceptedTransaction.go [hunk 19]
```diff
@@ -155,39 +215,39 @@ func (inTx *InterceptedTransaction) processFields(txBuff []byte) error {
 }
 
 // integrity checks for not nil fields and negative value
-func (inTx *InterceptedTransaction) integrity() error {
-	if inTx.tx.Signature == nil {
+func (inTx *InterceptedTransaction) integrity(tx *transaction.Transaction) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/interceptedTransaction.go:218]: integrity is used like everywhere in interceptors now.

File: process/transaction/shardProcess.go [hunk 20]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, err)
+	}
+	if !bytes.Equal(userTx.SndAddr, tx.RcvAddr) {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrInvalidAddressInRelayedTx)
+	}
+	if userTx.Value.Cmp(tx.Value) < 0 {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxValueHigherThenUserTxValue)
+	}
+
+	relayerGasLimit := txProc.economicsFee.ComputeGasLimit(tx)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:442]: done

File: process/transaction/shardProcess.go [hunk 21]
```diff
@@ -393,6 +409,229 @@ func (txProc *txProcessor) moveBalances(
...
+	err = txProc.accounts.SaveAccount(acntDst)
+	if err != nil {
+		return 0, err
+	}
+
+	return txProc.processUserTx(userTx, adrSrc, tx.Value, tx.Nonce, txHash)
+}
+
+func (txProc *txProcessor) processUserTx(
+	tx *transaction.Transaction,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/shardProcess.go:495]: done

File: process/smartContract/process.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/process.go:507]: done

**Pull request merged**